Add custom_range argument for partial dependence

freddyaboulton opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #20890 

#### What does this implement/fix? Explain your changes.
This PR allows users to specify a `custom_range` of values to calculate partial depedency for some or all of the `features`.

The api is `custom_range={feature: array-like of grid values}`. 

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

freddyaboulton committed:
Add custom_grid argument for partial dependence

freddyaboulton commented:
For whoever reviews this PR, where in the changelog I should add this change?

thomasjpfan commented:
> For whoever reviews this PR, where in the changelog I should add this change?

doc/whats_new/v1.1.rst

freddyaboulton committed:
Add 21033 to release notes

freddyaboulton committed:
Fix spacing

thomasjpfan commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -321,6 +370,11 @@ def partial_dependence(
 
+    custom_grid: dict
+        A dictionary mapping an element of `features` to an array of values where
+        the partial dependence should be calculated for that feature. The length
+        of `custom_grid` must match the length of `features`.
```
I think the requirement that `len(custom_grid) == len(features)` is not a great user experience. 

I was thinking of extending `_grid_from_X` to support `custom_grid`, and if `custom_grid` is provide for a given feature, then do minimal validation and add it to `values`.

thomasjpfan reviewed (commented):
Thank you for starting this PR @freddyaboulton !

I am open to a better name than `custom_grid`, the dictionary's value is more like a range.

freddyaboulton committed:
Rename custom_grid to custom_range. custom_range can now contain a subset of features.

freddyaboulton committed:
Rename custom_grid to custom_range in changelog

freddyaboulton committed:
Rename custom_grid to custom_range in tests

freddyaboulton commented:
@thomasjpfan Thank you for the feedback! I changed the name from `custom_grid` to `custom_range` and changed the implementation so that `custom_range` can include a subset of the `features`.

thomasjpfan reviewed (commented):
Thank you for the follow up! I made another pass through the PR.

freddyaboulton committed:
Use object only when necessary. Change name to custom_values

freddyaboulton committed:
Add comment

freddyaboulton committed:
Add newline to prevent sphinx warning

bchen1116 commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -58,6 +58,10 @@ def _grid_from_X(X, percentiles, grid_resolution):
 
+    custom_values: dict
+        Mapping from column index of X to an array-like of values where
+        the partial dependence should be calculated for that feature
```
nit: period here

jimbudarz replied:
"Mapping from column index of X" isn't accurate, it's actually mapping the element number from the list of user-provided features. Not sure if this will have downstream effects.
i.e. If user specifies that they want PDP for feature 5 only, the key generated in the `partial_dependence` function will assign this feature as 0, not 5.

freddyaboulton replied:
I think the docstring is accurate. At this point, `X` has been subset to the features in the `features` array with `_safe_indexing` and the `custom_values` array is modified so that it maps the column index in the subset array to the array of values.

bchen1116 commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -36,7 +36,7 @@
 
 
-def _grid_from_X(X, percentiles, grid_resolution):
+def _grid_from_X(X, percentiles, grid_resolution, custom_values):
```
is it possible to default `custom_values` as an empty dict? This avoids the case where a value needs to be passed in to this as an argument when it isn't necessary

jimbudarz replied:
I believe it needs to default to `None`, then be replaced by an empty dictionary if it's `None`.

bchen1116 reviewed (commented):
Really cool work! Left a question on whether `custom_values` can default to an empty dictionary

jimbudarz commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -107,6 +107,54 @@ def _grid_from_X(X, percentiles, grid_resolution):
+        if value.ndim != 1:
+            raise ValueError(
+                "Grid for feature {} is not a one-dimensional array. Got {} dimensions"
+                .format(feature, value.ndim)
```
It looks like `feature` is actually an index here, not a name, so users will not get the feature name in the error.

jimbudarz commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -81,30 +86,59 @@ def _grid_from_X(X, percentiles, grid_resolution):
 
     values = []
     for feature in range(X.shape[1]):
```
It would make sense to rename this as `feature_idx` since it's not the feature name.

jimbudarz commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -81,30 +86,59 @@ def _grid_from_X(X, percentiles, grid_resolution):
-                _safe_indexing(X, feature, axis=1), prob=percentiles, axis=0
-            )
-            if np.allclose(emp_percentiles[0], emp_percentiles[1]):
+        if feature in custom_values:
```
It would make sense to rename `feature` to `feature_idx` since it's not the feature name.

jimbudarz replied:
`custom_values` is also index-based, and perhaps the name should reflect that (since `custom_values` is also an argument in the `partial_dependence` function, and has a different structure.

InterferencePattern reviewed (changes requested):
This is a great feature you've added to PDP. Look forward to getting it nailed down and released!

jimbudarz commented:
I have another question- does this work with `PartialDependenceDisplay` object's required `deciles` attribute?

Commit:
Added custom_values to PartialDependenceDisplay plotting

freddyaboulton committed:
Linting

freddyaboulton committed:
Add space to fix docs

jimbudarz commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -79,30 +77,59 @@ def _grid_from_X(X, percentiles, grid_resolution):
-    return cartesian(values), values
+    # Create a place holder for the cartesian product of the individual grids.
+    shape = (len(v) for v in values)
+    ix = np.indices(shape)
```
It's probably a bad idea to create an array with a new dimension for every single feature, because this line will fail for >32 features.

```
Traceback (most recent call last):
  File ".../lib/python3.7/site-packages/sklearn/inspection/_partial_dependence.py", line 521, in partial_dependence
    custom_values_idx
  File ".../lib/python3.7/site-packages/sklearn/inspection/_partial_dependence.py", line 120, in _grid_from_X
    ix = np.indices(shape)
  File ".../lib/python3.7/site-packages/numpy/core/numeric.py", line 1777, in indices
    res = empty((N,)+dimensions, dtype=dtype)
ValueError: maximum supported dimension for an ndarray is 32, found 792
```

freddyaboulton replied:
You're correct but that limitation exists on the current main branch

```python
from sklearn.datasets import load_digits
from sklearn.ensemble import RandomForestClassifier
from sklearn.inspection import partial_dependence
import pytest

data = load_digits()
X = data.data
y = data.target

rf = RandomForestClassifier()
rf.fit(X, y)

with pytest.raises(ValueError, match="maximum supported dimension for an ndarray is 32, found 33"):
    partial_dependence(rf, X, features=list(range(32)))
```

So if we want to support more than 32-way partial dependence, that's a separate ticket.

freddyaboulton committed:
Fix docs

freddyaboulton committed:
Fix docs in grid_from_X

freddyaboulton committed:
Fix docstring

freddyaboulton committed:
Add plotting tests

freddyaboulton commented:
@thomasjpfan Can you please give this another look when you get a chance? Thanks!

thomasjpfan commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -79,30 +84,59 @@ def _grid_from_X(X, percentiles, grid_resolution):
+                    "Grid for feature {} is not a one-dimensional array. Got {}"
+                    " dimensions".format(feature, feature_range.ndim)
+                )
+            axis = feature_range
```
I'll prefer to error check before hand the loop since checking `custom_values` is more of an input validation: (It also reduces the code that is indented:

```python
custom_values = {k: np.asarray(v) for k, v in custom_values.items()}
if any(v.ndim != 1 for v in custom_values.values()):
    raise ValueError(...)

for feature in range(X.shape[1]):
    if feature in custom_values:
        axis = custom_values[feature]
    else:
        ...
```

In the erorr message, I think it's enough to say that `custom_values` contain non-1d data. The feature index does not correspond to the location of the original `X` anymore so it can be confusing.

Also, we can add a new test to `test_grid_from_X_error` to check the `custom_values` validation

thomasjpfan commented on sklearn/inspection/_partial_dependence.py:
```diff
@@ -457,8 +498,19 @@ def partial_dependence(
+    custom_values = custom_values or {}
+    if isinstance(features, (str, int)):
+        features = [features]
+    custom_values_idx = {
+        index: custom_values.get(feature)
+        for index, feature in enumerate(features)
+        if feature in custom_values
+    }
```
I think it's fairly unclear that `custom_values_idx` maps from the sliced `X`. I think we can make it more clear by doing:

```python
X_subset = _safe_indexing(X, features_indices, axis=1)

custom_values = custom_values or {}
custom_values_for_X_subset = {...}
grid, values = _grid_from_X(X_subset, ...)
```

thomasjpfan reviewed (commented):
Thanks for the update!

stephenpardy commented:
@freddyaboulton, @thomasjpfan I am also interested in this feature. There hasn't been any activity on this PR in over a year and I would be interested in taking it on.

stephenpardy commented:
@thomasjpfan can we switch the source to my branch - https://github.com/stephenpardy/scikit-learn/tree/20890-partial-dependence-custom-grid? I talked to Freddy and I will be taking over the PR - I will address the issues brought up in the last comments.

thomasjpfan commented:
@freddyaboulton Can you confirm that you are okay with @stephenpardy continuing the work on this PR?

@stephenpardy It's best to open another PR that references this one and states that it is superseding this PR. The new PR's opening commit should still link back to the original issue and describe its updates.

freddyaboulton commented:
Hi @thomasjpfan ! Yes @stephenpardy can take over
